src/flamenco/vm/fd_vm_private.h

```diff
@@ -184,16 +196,86 @@ fd_vm_mem_cfg( fd_vm_t * vm ) {
    hit loads and pretty good ILP.
 
    fd_vm_mem_haddr_fast is when the vaddr is for use when it is already
-   known that the vaddr region has a valid mapping. */
+   known that the vaddr region has a valid mapping.
+   
+   These assumptions don't hold if direct mapping is enabled since input
+   region lookups become O(log(n)). */
+
+static inline ulong
+fd_vm_get_input_mem_region_idx( fd_vm_t const * vm, ulong offset ) {
+  uint left  = 0U;
+  uint right = vm->input_mem_regions_cnt - 1U;
+  uint mid   = 0U;
+
+  while( left<right ) {
+    mid = (left+right) / 2U;
+    if( offset>=vm->input_mem_regions[mid].vaddr_offset ) {
+      /* Exit condition: found the right region */
+      if( mid+1==vm->input_mem_regions_cnt || offset<vm->input_mem_regions[mid+1].vaddr_offset ) {
+        break;
+      }
+      left = mid + 1U;
+    } else {
+      right = mid - 1U;
+    }
+  }
+  if( left==right ) {
+    mid = left;
+  }
+  left = mid;
+  return left;
```

Comment:
I think we can just return mid here, no need for these lines of code.

Reply:
isn't the same because of exit conditions. Will clean up though